### PR TITLE
Add occ preview:migrate to migrate previews from the old flat structure to a subfolder structure

### PR DESCRIPTION
**Pull request by MorrisJobke**
## What was added?

* `php occ preview:repair` - a preview migration tool that moves existing previews into the new location introduced with #19214
* moves `appdata_INSTANCEID/previews/FILEID` to `appdata_INSTANCEID/previews/0/5/8/4/c/e/5/FILEID`
* migration tool can be stopped during migration via `CTRL+C` - it then finishes the current folder (with the previews of one file) and stops gracefully
* if a PHP memory limit is set in the `php.ini` then it will stop automatically once it has less than 25 MiB memory left (this is to avoid hard crashes in the middle of a migration)
* the tool can be used during operation - possible drawbacks:
    * there is the chance of a race condition that a new preview is generated in the moment the folder is already migrated away - so the old folder with the newly cached preview is deleted and one cached preview needs to be re-generated
    * there is the chance of a race condition during access of a preview while it is migrated to the other folder - then no preview can be shown and results in a 404 (as of now this is an accepted risk)

## How does it look like?

Bildschirmfoto 2020-08-06 um 22 04 50

## How to test?

* Do this change:

```patch
diff --git a/lib/private/Preview/Storage/Root.php b/lib/private/Preview/Storage/Root.php
index 1f332c3208..b6de590dc9 100644
--- a/lib/private/Preview/Storage/Root.php
+++ b/lib/private/Preview/Storage/Root.php
@@ -55,7 +55,7 @@ class Root extends AppData {
 
 	public function newFolder(string $name): ISimpleFolder {
 		$internalFolder = self::getInternalFolder($name);
-		return parent::newFolder($internalFolder);
+		return parent::newFolder($name);
 	}
 
 	/*
```

* Upload a new file that generates previews (i.e. an image)
* Now previews are generated in the old flat folder structure `appdata_INSTANCEID/previews/FILEID`
* Run `occ preview:repair` (optional with `-d` and/or `-v`) to check the operations out 

Ref #22063 and #22033 (it originated there but this is the essence and might be useful for all setups).

### CODE REVIEW

File: core/Command/Preview/Repair.php [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+		$currentPreviewFolder = $this->rootFolder->get("appdata_$instanceId/preview");
+
+		$directoryListing = $currentPreviewFolder->getDirectoryListing();
+
+		$total = count($directoryListing);
+		/**
+		 * by default there could be 0-9 a-f and the old-multibucket folder which are all fine
+		 */
+		if ($total < 18) {
+			foreach ($directoryListing as $index => $dir) {
```
Comment (icewind1991) [inline core/Command/Preview/Repair.php:115]: `array_filter` would be much clearer here imo

**Review by icewind1991 — APPROVED** [review]
Looks good besides the minor style complaint

**Review by juliusknorr — APPROVED** [review]
Tested and works and code looks good :+1:

File: core/Command/Preview/Repair.php (hunk 1 above)
Comment (MorrisJobke) [reply core/Command/Preview/Repair.php]: I added a test case that verifies the current behavior and then changed this into an `array_filter`.

**Review by juliusknorr — APPROVED** [review]

**Comment by agross** [comment]
What unfortunately wasn't considered here is that the repair could be interrupted by e.g. shutting down the server as well. For my instance this resulted in ~1.5 million uncommitted rows that MySQL will now roll back. At 500 rows/15 seconds this will take forever. Almost as much as converting to the new directory structure, which was projected to take 14 days.

**Pull request merged**